File: longestHappyString1405.cpp
class Solution {
private:
    char characts[3] = {'a', 'b', 'c'};
    char getMax(int* arr, int expt) {
        int idx = -1, max = INT_MIN;
        if (expt != 0 && arr[0] > max) {
            idx = 0;
            max = arr[0];
        }
        if (expt != 1 && arr[1] > max) {
            idx = 1;
            max = arr[1];
        }
        if (expt != 2 && arr[2] > max) {
            idx = 2;
            max = arr[2];
        }
        if (max == 0) return ' ';
        arr[idx]--;
        return characts[idx];
    }
public:
    string longestDiverseString(int a, int b, int c) {
        int arr[3] = {a, b, c};
        string ans;
        ans = ans + getMax(arr, -1);
        ans = ans + getMax(arr, -1);
        while (true) {
            int size = ans.size();
            int expt = -1;
            if (ans[size - 1] == ans[size - 2]) {
                expt = ans[size - 1] - 'a';
            }
            char next = getMax(arr, expt);
            if (next == ' ') {
                break;
            }
            ans = ans + next;
        }
        return ans;
    }
};

File: longestHappyString1405.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        sum = a+b+c
        A, B, C = 0, 0, 0
        s = []
        ans = ""
        for i in range(0, sum):
            if (a>=b and a>= c and A !=2) or (B == 2 and a > 0) or (C == 2 and a > 0):
                s.append('a')
                a-=1
                A+=1
                B, C = 0,0
            elif (b>=a and b>= c and B !=2) or (A == 2 and b > 0) or (C == 2 and b > 0):
                s.append('b')
                b-=1
                B+=1
                A, C= 0,0
            elif (c>=a and c>= b and C !=2) or (A == 2 and c > 0) or (B == 2 and c > 0):
                s.append('c')
                c-=1
                C+=1
                A=B=0
        for c in s:
            ans+=c 
        return ans 
